lib/internal/test_runner/runner.js

```diff
@@ -416,7 +416,22 @@ function watchFiles(testFiles, opts) {
   const watcher = new FilesWatcher({ __proto__: null, debounce: 200, mode: 'filter', signal: opts.signal });
   const filesWatcher = { __proto__: null, watcher, runningProcesses, runningSubtests };
   opts.root.harness.watching = true;
+  // Watch for created/deleted files in the current directory
+  const allFilesWatcher = new FilesWatcher({ __proto__: null, debounce: 200, mode: 'all', signal: opts.signal });
+  allFilesWatcher.watchPath(process.cwd());
+  allFilesWatcher.on('changed', ({ owners, eventType }) => {
+    if (eventType === 'rename') {
+      const updatedTestFiles = createTestFileList(opts.globPatterns);
+
+      const newFiles = ArrayPrototypeFilter(updatedTestFiles, (x) => !ArrayPrototypeIncludes(testFiles, x));
 
+      testFiles = updatedTestFiles;
+      newFiles.forEach((newFile) => {
+        watcher.emit('changed', { __proto__: null, owners: new SafeSet().add(newFile), eventType: 'rename' });
+      });
```

Comment:
I don't understand this, why?

Reply:
The other watcher is in `filter` mode, and a lot of logic revolves around the filter mode in the lines below.
To avoid changing the current implementation, I tried to add another watcher set to `all` mode, intended to intercept every change in the directory and its subfolders.
As soon as a `rename` occurs, I'm emitting a `changed` `rename` event to the other watcher to allow it to execute the logic we currently use to spin up a new test (e.g., during a filtered file rename or update).

This piece of code smells to me, but I couldn't come up with a better solution.
It works, but it seems like a very poor approach.

Do you have any suggestions?

Reply:
@cjihrig have you got some suggestion for this?

Reply:
I removed the unnecessary additional watcher and changed the original watcher mode to `all`.

In order to prevent the runner from returning immediately, I had to modify `watchChildProcessModules` (lib/internal/watch_mode/files_watcher.js) by removing:
```js
    if (this.#mode !== 'filter') {
      return;
    }
```

I left a comment in the code about this since I'm unable to figure out why it was there in the first place.

Currently, I've noticed that tests are failing in a Linux environment, and I suspect the issue might be related to differences in the `rename` and `change` events between environments.
(This issue doesn't occur on macOS)

Reply:
With my last commit, I fixed the failing tests in Linux.
I would say that the implementation is ready to be reviewed again